### PR TITLE
sci-libs/spyking-circus: version bump and fix dependencies

### PR DESCRIPTION
**Pull request by malfatti**
Hi!

This ebuild fixes dependencies, specially releated to mpi. The previous version may not work because spyking-circus depends on specific versions of `openmpi` or `mpich`. Also, bump.

Cheers :)

### CODE REVIEW

**Comment by TheChymera** [comment]
@malfatti hey, thank you for this!

I can help you merge this, but dependency resolution for this package seems to break on my end:

```
silenthost ~ # emerge -av spyking-circus

These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  N     ] dev-python/colorama-0.4.3::gentoo  USE="-examples -test" PYTHON_TARGETS="python2_7 python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 0 KiB
[ebuild  N     ] dev-python/tqdm-4.48.2::gentoo  USE="-examples -test" PYTHON_TARGETS="python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 146 KiB
[ebuild  N     ] dev-python/distro-1.5.0::gentoo  USE="-test" PYTHON_TARGETS="python2_7 python3_7 -python3_6 -python3_8 -python3_9" 0 KiB
[ebuild  N     ] app-arch/lz4-1.9.2:0/r132::gentoo  USE="-static-libs" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] app-arch/snappy-1.1.8:0/1::gentoo  USE="-test" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] sys-apps/hwloc-2.1.0:0/15::gentoo  USE="X cairo cpuid pci svg xml -cuda -debug -gl -libudev -netloc -nvml -plugins -static-libs" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] dev-libs/c-blosc-1.20.1:0/1::gentoo  USE="lz4 snappy zlib -test -zstd" 859 KiB
[ebuild  N     ] dev-python/scikit-build-0.11.1::gentoo  USE="-doc -test" PYTHON_TARGETS="python3_7 -python3_6 -python3_8 -python3_9" 0 KiB
[ebuild  N     ] dev-python/blosc-1.9.2::gentoo  USE="-test" PYTHON_TARGETS="python3_7 -python3_6 -python3_8 -python3_9" 798 KiB
[ebuild  N     ] sys-cluster/openmpi-4.0.4::gentoo  USE="fortran ipv6 -cma -cuda -cxx -heterogeneous -java -romio" ABI_X86="(64) -32 (-x32)" OPENMPI_FABRICS="-knem -ofed -psm" OPENMPI_OFED_FEATURES="-control-hdr-padding -dynamic-sl -rdmacm -udcm" OPENMPI_RM="-pbs -slurm" 9,520 KiB
[ebuild  N     ] sys-cluster/mpich-3.3::gentoo  USE="cxx fortran -doc -mpi-threads -romio -threads" ABI_X86="(64) -32 (-x32)" 26,572 KiB
[ebuild  N     ] virtual/mpi-2.0-r5::gentoo  USE="fortran -cxx -romio -threads" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] dev-python/mpi4py-3.0.3::gentoo  USE="-doc -examples -test" PYTHON_TARGETS="python3_7 -python3_6 -python3_8 -python3_9" 0 KiB
[ebuild  N     ] sci-libs/spyking-circus-0.9.9::chymeric  PYTHON_TARGETS="python3_7 -python3_6" 4,662 KiB
[blocks B      ] sys-cluster/mpich ("sys-cluster/mpich" is blocking sys-cluster/openmpi-4.0.4)
[blocks B      ] sys-cluster/openmpi ("sys-cluster/openmpi" is blocking sys-cluster/mpich-3.3)

Total: 14 packages (14 new), Size of downloads: 42,555 KiB
Conflict: 2 blocks (2 unsatisfied)

 * Error: The above package list contains packages which cannot be
 * installed at the same time on the same system.

  (sys-cluster/mpich-3.3:0/0::gentoo, ebuild scheduled for merge) pulled in by
    >=sys-cluster/mpich-2.0 required by (sci-libs/spyking-circus-0.9.9:0/0::chymeric, ebuild scheduled for merge) USE="" ABI_X86="(64)" PYTHON_TARGETS="python3_7 -python3_6"

  (sys-cluster/openmpi-4.0.4:0/0::gentoo, ebuild scheduled for merge) pulled in by
    >=sys-cluster/openmpi-1.10.2-r1[abi_x86_32(-)?,abi_x86_64(-)?,abi_x86_x32(-)?,abi_mips_n32(-)?,abi_mips_n64(-)?,abi_mips_o32(-)?,abi_riscv_lp64d(-)?,abi_riscv_lp64(-)?,abi_riscv_ilp32d(-)?,abi_riscv_ilp32(-)?,abi_s390_32(-)?,abi_s390_64(-)?,cxx?,fortran?,romio?,threads(+)?] (>=sys-cluster/openmpi-1.10.2-r1[abi_x86_64(-),fortran]) required by (virtual/mpi-2.0-r5:0/0::gentoo, ebuild scheduled for merge) USE="fortran -cxx -romio -threads" ABI_X86="(64) -32 (-x32)"


For more information about Blocked Packages, please refer to the following
section of the Gentoo Linux x86 Handbook (architecture is irrelevant):

https://wiki.gentoo.org/wiki/Handbook:X86/Working/Portage#Blocked_packages
```
Something that might be relevant to this issue: In my experience many python packages are overly strict in their dependency specification (i.e. the developer just puts whatever dependency version he had as the only one that's compatible). This is almost always not a true constraint, and can break stuff for systems which have proper dependency management instead of just relying on stuff hacked together in a virtualenv :)

Also, remember to run the test suite, to check that it actually works, and if you find the time, remind upstream that it would be really cool if their test suite worked without qt :)
https://github.com/spyking-circus/spyking-circus/issues/234

**Comment by malfatti** [comment]
Hey @TheChymera,

Oops, I guess since I already have openmpi installed it didn't complain :) 

Actually I would appreciate your help in solving this particular dependency. My problem is with the syntax `|| ( <=sys-cluster/openmpi-3.0.0 >=sys-cluster/mpich-2.0 )`. I thought it would be satisfied with one OR the other (because the two mpi implementations are mutually exclusive), but somehow it is trying to emerge openmpi-4.0.4 and mpich-3.3 :/  

About the over-specific versions, I'm not sure about the other dependencies (I added the versions that pip tries to install), but the mpi implementation is even a pinned issue upstream ([spyking-circus/spyking-circus#93](https://github.com/spyking-circus/spyking-circus/issues/93)). It has to be `<=sys-cluster/openmpi-3.0.0` or `>=sys-cluster/mpich-2.0`.

Thanks for looking into this :)

**Comment by TheChymera** [comment]
@malfatti hm, I have no idea. Can you not reproduce this issue locally?

**Comment by TheChymera** [comment]
@malfatti the error apparently results from the fact that my newer `=sys-cluster/openmpi-4.0.4` does not satisfy the `<=sys-cluster/openmpi-3.0.0` constraint. For some reason portage does not think of downgrading openmpi and instead goes for the alternative. Why downgrade is not the preferred option is an interesting question which you could clarify with #gentoo-dev-help or #gentoo-portage on freenode IRC (easy to join, lots of helpful people).

In any case the package builds with 

```
|| ( sys-cluster/openmpi >=sys-cluster/mpich-2.0 )
```

so it might be worth considering whether this was an artificial constraint, as mentioned above. To better test proper functionality in this case, would you have time to please enable the test suite?

**Comment by malfatti** [comment]
Hey @TheChymera,

I tried it in a previous version (0.8.6) and it was not artificial. It compiles and installs fine, but all parallel jobs fail (and parallel jobs are the main advantage of this spk clustering software :) ). Sorry I went silent for so long, as soon as possible I'll retest if this new version works with newer openmpi, and try to enable the test suite (never done it before :P )

Cheers

**Comment by TheChymera** [comment]
@malfatti let me know if you have any questions. You can also join us on #gentoo-science on freenode or #gentoo-dev-help for ebuild-writing-related questions.

**Comment by malfatti** [comment]
Sorry for the delay. I tested the package without the `<=sys-cluster/openmpi-3.0.0` constraint, and indeed it merges fine. However, when trying to run I got:
```bash
----------------------------  Error  -----------------------------
| SpyKING CIRCUS does not work with OPENMPI >= 3.0
| Consider downgrading or switching to MPICH
------------------------------------------------------------------
```

Manually trying to downgrade to `=sys-cluster/openmpi-2.1.6` made emerge to try to downgrade `hwloc` to `=sys-apps/hwloc-1.11.13`, which failed compiling:
```bash
[...]
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c: In function ‘hwloc_opencl_query_devices’:
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:108:5: error: unknown type name ‘cl_device_topology_amd’
  108 |     cl_device_topology_amd amdtopo;
      |     ^~~~~~~~~~~~~~~~~~~~~~
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:171:9: error: ‘CL_DEVICE_TOPOLOGY_TYPE_PCIE_AMD’ undeclared (first use in this function); did you mean ‘CL_DEVICE_TOPOLOGY_AMD’?
  171 |     if (CL_DEVICE_TOPOLOGY_TYPE_PCIE_AMD != amdtopo.raw.type) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         CL_DEVICE_TOPOLOGY_AMD
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:171:9: note: each undeclared identifier is reported only once for each function it appears in
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:171:52: error: request for member ‘raw’ in something not a structure or union
  171 |     if (CL_DEVICE_TOPOLOGY_TYPE_PCIE_AMD != amdtopo.raw.type) {
      |                                                    ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:172:53: error: request for member ‘raw’ in something not a structure or union
  172 |       hwloc_debug("not a PCIe device: %u\n", amdtopo.raw.type);
      |                                                     ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:178:40: error: request for member ‘pcie’ in something not a structure or union
  178 |     info->specific.amd.pcibus = amdtopo.pcie.bus;
      |                                        ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:179:40: error: request for member ‘pcie’ in something not a structure or union
  179 |     info->specific.amd.pcidev = amdtopo.pcie.device;
      |                                        ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:180:41: error: request for member ‘pcie’ in something not a structure or union
  180 |     info->specific.amd.pcifunc = amdtopo.pcie.function;
      |                                         ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:183:21: error: request for member ‘pcie’ in something not a structure or union
  183 |   (unsigned) amdtopo.pcie.bus, (unsigned) amdtopo.pcie.device, (unsigned) amdtopo.pcie.function);
      |                     ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:183:50: error: request for member ‘pcie’ in something not a structure or union
  183 |   (unsigned) amdtopo.pcie.bus, (unsigned) amdtopo.pcie.device, (unsigned) amdtopo.pcie.function);
      |                                                  ^
/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13/src/topology-opencl.c:183:82: error: request for member ‘pcie’ in something not a structure or union
  183 |   (unsigned) amdtopo.pcie.bus, (unsigned) amdtopo.pcie.device, (unsigned) amdtopo.pcie.function);
      |                                                                                  ^
make[1]: *** [Makefile:923: topology-opencl.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/var/tmp/portage/sys-apps/hwloc-1.11.13/work/hwloc-1.11.13-abi_x86_64.amd64/src'
make: *** [Makefile:605: all-recursive] Error 1
```

Also, switching to mpich is not an option in practice, since apparently several packages (including `dev-libs/boost`) depend on `virtual/mpi`, which pulls `sys-cluster/openmpi`.

The only way I see is to have multiple mpi implementations, but I have no Idea how to implement this :) Or maybe we can wait for the openmpi constraint to be fixed upstream...

**Comment by TheChymera** [comment]
@malfatti hm, that is sad. Can you create an issue upstream asking them to do this, and link it here?

For the time being, we could just merge it with the original constraint. The only problem seemed to be that if you have a newer openmpi you need to manually downgrade. This is just a portage issue. It might be possible to reformulate the constraint so that portage tries to downgrade first. Can you check with #gentoo-dev-help?

Also, is there a test phase for the package you can add, so that we can be sure that whenever we relax the dependency constraint it still works?

**Comment by epsilon-0** [comment]
MPI subsytem in gentoo is a mess. It's even worse than BLAS/LAPACK. No one has a clean solution for MPI.
There is also no slotted MPI available.

A possible solution - add a blocker for openmpi, this way when someone tries to pull in virtual/mpi they will be forced to use another MPI implementation (mpich is the only other "good" one).
But I agree with @TheChymera it would be best to go upstream and ask them to change :broken_heart:

**Comment by malfatti** [comment]
Hey @TheChymera and @epsilon-0,

There is already an (pinned, even) issue about mpi upstream ([spyking-circus/spyking-circus#93](https://github.com/spyking-circus/spyking-circus/issues/93)), opened in 2018 with the last comment on 08/2019 :/

I just followed the blocker approach. Now it depends on `>=sys-cluster/mpich-2.0`, and since it already has a `sys-cluster/openmpi` blocker, this constraint should be "solved" (users without `sys-cluster/openmpi` will happily merge and run the software).

About the test suite, I spent the afternoon playing with it and... nothing. The `tests/__init__.py` is written in python2. I tried writing some patches at first, without success. Then I tried adding
```bash
python_prepare() {
    if python_is_python3; then
        2to3 -n -w --no-diffs tests/__init__.py || die
    fi
}
```
but there were still unavailable libraries (cPickle, urllib2...) being requested. I gave up. Sorry :P

**Comment by epsilon-0** [comment]
OK, I agree with dropping tests, this is too painful.

@TheChymera can do the honors. :smile_cat:

**Comment by TheChymera** [comment]
@malfatti can you please reintroduce the correct dependency constraint, and rebase everything into one commit for the new package?

**Comment by malfatti** [comment]
@TheChymera Done :)

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+inherit distutils-r1
+
+DESCRIPTION="Fast spike sorting by template matching"
+HOMEPAGE="https://github.com/spyking-circus/spyking-circus/"
+SRC_URI="https://github.com/spyking-circus/spyking-circus/archive/${PV}.tar.gz"
+LICENSE="CeCILL-2"
+
+SLOT="0"
+KEYWORDS="~amd64 ~amd64-linux ~x86 ~x86-linux"
```
Comment (TheChymera) [inline sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:16]: Please sort the keywords

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	>=dev-python/pandas-0.21[${PYTHON_USEDEP}]
+	>=dev-python/psutil-5.6.7[${PYTHON_USEDEP}]
+	>=dev-python/statsmodels-0.10.1[${PYTHON_USEDEP}]
+	>=dev-python/tqdm-4.40.0[${PYTHON_USEDEP}]
+	>=dev-python/scipy-1.3.1[${PYTHON_USEDEP}]
+	>=sci-libs/scikits_learn-0.21.3
+	|| ( <=sys-cluster/openmpi-3.0.0 >=sys-cluster/mpich-2.0 )
+"
+DEPEND="${RDEPEND}"
+
```
Comment (TheChymera) [inline sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:35]: Blank line here not needed

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	>=dev-python/h5py-2.9.0[${PYTHON_USEDEP}]
+	>=dev-python/matplotlib-2.2.4[${PYTHON_USEDEP}]
+	>=dev-python/mpi4py-3.0.0[${PYTHON_USEDEP}]
+	>=dev-python/numpy-1.17.4[${PYTHON_USEDEP}]
+	>=dev-python/pandas-0.21[${PYTHON_USEDEP}]
+	>=dev-python/psutil-5.6.7[${PYTHON_USEDEP}]
+	>=dev-python/statsmodels-0.10.1[${PYTHON_USEDEP}]
+	>=dev-python/tqdm-4.40.0[${PYTHON_USEDEP}]
+	>=dev-python/scipy-1.3.1[${PYTHON_USEDEP}]
+	>=sci-libs/scikits_learn-0.21.3
```
Comment (TheChymera) [inline sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:31]: New name of dependency is `sci-libs/scikit-learn`

**Review by TheChymera — CHANGES_REQUESTED** [review]
@malfatti Hey man, sorry again for missing some points in the review. As a more general point, please remember to run `repoman full` before committing, it helps point out such issues to you :)

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild (hunk 1 above)
Comment (malfatti) [reply sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:16]: I didn't get it:
```bash
$ echo "~amd64 ~amd64-linux ~x86 ~x86-linux" | tr ' ' '\n' | sort | tr '\n' ' '
~amd64 ~amd64-linux ~x86 ~x86-linux
```
Aren't they already sorted?

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild (hunk 3 above)
Comment (malfatti) [reply sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:31]: Changed :)

File: sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild (hunk 2 above)
Comment (malfatti) [reply sci-libs/spyking-circus/spyking-circus-0.9.9.ebuild:35]: Removed :)

**Comment by malfatti** [comment]
Sorry for missing the `repoman full` every time :P Thanks a lot for your patience :) Fixed everything and rebased.

**Comment by malfatti** [comment]
Hi!

Any clue on why this was closed?

**Comment by TheChymera** [comment]
@malfatti it was closed automatically by me appending a line saying `Closes: https://github.com/gentoo/sci/pull/965` in the respective commit. Gentoo science is in fact not primarily hosted on GH, there's no “merged” status for PRs: https://github.com/gentoo/sci/pulls?q=is%3Apr+is%3Aclosed All PRs get closed after the author's commit is merged.

**Comment by malfatti** [comment]
Ah cool :) Thanks!

**Pull request closed without merging**